pennylane/measurements/state.py

```diff
@@ -116,4 +119,18 @@ def circuit():
     """
     # pylint: disable=protected-access
     wires = Wires(wires)
-    return MeasurementProcess(State, wires=wires)
+    return _State(State, wires=wires)
+
+
+class _State(StateMeasurement):
+    """Measurement process that returns the quantum state."""
+
+    # pylint: disable=redefined-outer-name
+    def process_state(self, state: Sequence[complex], wires: Wires):
+        if self.wires == Wires([]):
+            # qml.state
+            return state
+        # qml.density_matrix
+        wire_map = dict(zip(range(len(wires)), wires))
+        mapped_wires = [wire_map[w] for w in self.wires]
+        return qml.math.reduced_dm(state, indices=mapped_wires, c_dtype="complex128")
```

Comment:
If I understand correctly, `wires` represents the device wires and `self.wires` represent the wires of the measurement? In that case, it seems like this wouldn't work for non-integer wire labels in the device.

Reply:
Yes! `wires` are the wires corresponding to the `state` (which are the device wires).

Reply:
You are right! The mapping has to be swapped. Changed it to:
```python
wire_map = dict(zip(wires, range(len(wires))))
```

Also added test to make sure it works.